### ISSUE TITLE
Wayland/X11 support?

### ISSUE THREAD

**Issue by tommythorn** [rbreaves/kinto#420]
**Describe the bug**
Running a essentially stock Ubuntu 21.04 (beta), kinto's `setup.py` complains
```
Failed to get session path: Caller does not belong to any known session.
You are not using x11 [sic], please logout and back in using x11[sic]/Xorg
```
(It's X11)

Unfortunately, there's no "cog wheel" in my login screen so there's no immediately
obvious way to do that.

If I understand it correctly the issue is that Kinto doesn't work with the Wayland X server,
however it seems Ubuntu is increasingly deprecating the Xorg option, so Kinto will
thus cease to be an option.

**Expected behavior**
Well, obviously, work.

**Install Type:** Bare Metal
**Distro:** Ubuntu 21.04 (Hirsute Hippo) (beta, but fully up-to-date)
**DE:** Gnome (it seems)
**Branch:** master
**Commit:** 4980c4a35ea7735e4f4a1e8a2f6e45b6d1ea7603

Logs and status if relevant
N/A

**Comment by rbreaves** [comment]
The work to bring Wayland compatibility to Kinto at the forefront of my mind - but it will need to be added to xkeysnail as that is what Kinto uses to handle remappings, my own fork of xkeysnail in actuality.

If users would like to engage with that conversation on over in this thread https://github.com/mooz/xkeysnail/issues/108 then I will be more than happy to collaborate with others to bring Wayland support sooner rather than later. It will be more piecemeal, dependent on the DE you use rather than Wayland itself though, so I would expect to be targeting the most popular DEs that support Wayland first.

**Comment by xpusostomos** [comment]
@tommythorn At least for me... the cog wheel appears after you click on the login field.

**Comment by njelich** [comment]
Do you think this would be useful? https://github.com/arnarg/waybind

**Comment by rbreaves** [comment]
> Do you think this would be useful? https://github.com/arnarg/waybind or https://github.com/snyball/Hawck

Not really, spoke w/ at least one of the authors already & it’s the same issue we all have - there’s no universal way to query the active app under Wayland & no one wants to take the time to write a wrapper or something to interface w/ all the DEs to get the active app.

Kinto via xkeysnail remaps just fine on Wayland - but it can’t dynamically change the keymap based on the app or app type until someone fixes the app detection.

**Comment by Atemu** [comment]
It doesn't seem like there is a native wayland method of achieving this.

Perhaps you could use a low-lever remapper in a dynamic manner where a daemon would listen to [wlr-foreign-toplevel-management](https://wayland.app/protocols/wlr-foreign-toplevel-management-unstable-v1) and use IPC to dictate key mappings to the low-level remapper dynamically.  
I don't know whether any one is capable of being configured dynamically via IPC but it doesn't sound impossible and doesn't require standards.

There would be a very slight delay after switching I'd imagine but I highly doubt you could notice.

**Comment by geofflittle** [comment]
I'm on Ubuntu 22.04 and following the install instructions at [1], repro'd here:

```
/bin/bash -c "$(wget -qO- https://raw.githubusercontent.com/rbreaves/kinto/HEAD/install/linux.sh || curl -fsSL https://raw.githubusercontent.com/rbreaves/kinto/HEAD/install/linux.sh)"
```

When I paste the above command into my terminal I receive the error

```
...
Failed to get session path: Caller does not belong to any known session.
You are not using x11, please logout and back in using x11/Xorg
```

[1] - Quick install instructions: https://github.com/rbreaves/kinto#quick-install-method

**Comment by rbreaves** [comment]
When you logout there will be a small gear icon in the bottom right of the login screen - switch to x11/xorg there.

**Comment by geofflittle** [comment]
@rbreaves Selecting x11 from the gear menu at login solved my issue.  Thanks!

**Comment by joshgoebel** [comment]
> where a daemon would listen to wlr-foreign-toplevel-management 

This is interesting, but it seems to only talk about visibility - NOT input focus...

> This event is emitted whenever the toplevel becomes visible on the given output.

Unless there is a lot more implied by "toplevel"...

**Comment by cphoward** [comment]
@geofflittle Were you able to get it installed? I've encountered the same problem on Ubuntu 22.04.3 LTS.

**Comment by hexplor** [comment]
At some point, when I launch and configure everything under Ubuntu 24.04.3 LTS X11, then switch to Wayland, everything seems to be working. All my remaps, keys, and tray icon.

**Comment by RedBearAK** [comment]
@hexplor 

> At some point, when I launch and configure everything under Ubuntu 24.04.3 LTS X11, then switch to Wayland, everything seems to be working. All my remaps, keys, and tray icon.

Only general (not app-specific) remaps will work in Wayland, because there is no window context for the keymapper to use to enable any of the app-specific keymaps. There will not be proper app-specific remapping of shortcuts in terminals, or file managers, or web browsers, or code editors like VSCode. 

This will be most immediately obvious in terminals, where Cmd+C will do a SIGINT and terminate running processes, instead of copying the selected text. Normally Cmd+C is remapped to Shift+Ctrl+C in terminals to prevent the SIGINT, and many other terminal remaps do the same, remapping Ctrl to Shift+Ctrl. That can't happen without the window context info. But there will be many other applications affected, once you use something other than Cmd+Z/X/C/V and similar simple global remaps. 

If you want a shortcut remapping experience like Kinto that actually works in Wayland, you will have to switch to using Toshy, the alternative to Kinto that I started building about four years ago, mostly because of the lack of Wayland support in Kinto. It's got all the Kinto remaps and various kinds of improvements to the Kinto config, plus support for several different types of Wayland environments hacked into a fork of the keymapper. 

Toshy's installer also supports 16 different basic Linux distro types, covering most desktop distros in existence. Not all distros, but many more than Kinto's setup script supports. And it uses a Python virtual environment to work better in "managed" Python environments like recent Ubuntu releases. So it's abnormal for there to be any problems installing on common distros. 

If anyone wants to try Toshy while having Kinto installed, you'll need to stop Kinto from its tray or GUI app menu, and disable the auto-start, so that its keymapper service will not conflict with the Toshy keymapper service. 

https://github.com/RedBearAK/Toshy (short redirect URL https://toshy.app)

GNOME users need to look at Requirements in the Toshy README to get a shell extension installed that will provide the Wayland window context. Toshy handles all other Wayland environments automatically.
